pkg/reconciler/pubsub/topic/topic.go

```diff
@@ -0,0 +1,441 @@
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		logging.FromContext(ctx).Desugar().Info("Topic became ready", zap.Any("after", duration))
+		r.Recorder.Event(topic, corev1.EventTypeNormal, "ReadinessChanged", fmt.Sprintf("Topic %q became ready", topic.Name))
+		if err := r.StatsReporter.ReportReady("Topic", topic.Namespace, topic.Name, duration); err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to record ready for Topic", zap.Error(err))
+		}
+	}
+
+	return ch, err
+}
+
+// updateFinalizers is a generic method for future compatibility with a
+// reconciler SDK.
+func (r *Reconciler) updateFinalizers(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, bool, error) {
+	source, err := r.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Don't modify the informers copy.
+	existing := source.DeepCopy()
+
+	var finalizers []string
+
+	// If there's nothing to update, just return.
+	existingFinalizers := sets.NewString(existing.Finalizers...)
+	desiredFinalizers := sets.NewString(desired.Finalizers...)
+
+	if desiredFinalizers.Has(finalizerName) {
+		if existingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Add the finalizer.
+		finalizers = append(existing.Finalizers, finalizerName)
+	} else {
+		if !existingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return desired, false, nil
+		}
+		// Remove the finalizer.
+		existingFinalizers.Delete(finalizerName)
+		finalizers = existingFinalizers.List()
+	}
+
+	mergePatch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers":      finalizers,
+			"resourceVersion": existing.ResourceVersion,
+		},
+	}
+
+	patch, err := json.Marshal(mergePatch)
+	if err != nil {
+		return desired, false, err
+	}
+
+	update, err := r.RunClientSet.PubsubV1alpha1().Topics(existing.Namespace).Patch(existing.Name, types.MergePatchType, patch)
+	return update, true, err
+}
+
+func addFinalizer(s *v1alpha1.Topic) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Insert(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func removeFinalizer(s *v1alpha1.Topic) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Delete(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *Reconciler) reconcilePublisher(ctx context.Context, topic *v1alpha1.Topic) (error, *servingv1.Service) {
+	name := resources.GeneratePublisherName(topic)
+	existing, err := r.serviceLister.Services(topic.Namespace).Get(name)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			logging.FromContext(ctx).Desugar().Error("Unable to get an existing publisher", zap.Error(err))
+			return err, nil
+		}
+		existing = nil
+	} else if !metav1.IsControlledBy(existing, topic) {
+		p, _ := json.Marshal(existing)
+		logging.FromContext(ctx).Desugar().Error("Got a pre-owned publisher", zap.Any("publisher", p))
```

Comment:
This error message is oddly worded. I suggest changing it to be similar to the message below: `Topic does not own publisher service`

Reply:
Done, thanks!